### PR TITLE
Federated Server functions

### PR DESCRIPTION
**Pull request by Algunenano**
Add some new functions to manage a federated server.

This includes parts of the code from https://github.com/CartoDB/cartodb-postgresql/pull/373 and https://github.com/CartoDB/cartodb-postgresql/pull/381 and probably deprecates code recently added to CDB_ForeignTable.

As a WIP it still misses quite a bit of tests around corner cases (server / schema not existing, odd names...) and has some TODO's left (permissions, possible improvements) but it can be reviewed so we confirm this is the API we want to use.

Public functions:
* CDB_Federated_Server_List_Servers(server TEXT DEFAULT '%')
* CDB_Federated_Server_Register_PG(server TEXT, config JSONB)
* CDB_Federated_Server_Unregister(server TEXT)
* CDB_Federated_Server_List_Remote_Schemas(server TEXT)
* CDB_Federated_Server_List_Remote_Tables(server TEXT, remote_schema TEXT)
* CDB_Federated_Table_Register(
    server TEXT,
    remote_schema TEXT,
    remote_table TEXT,
    id_column TEXT,
    geom_column TEXT DEFAULT NULL,
    webmercator_column TEXT DEFAULT NULL,
    local_name NAME DEFAULT NULL
)
* CDB_Federated_Table_Unregister(
    server TEXT,
    remote_schema TEXT,
    remote_table TEXT
)
* CDB_Federated_Server_List_Registered_Tables(
    server TEXT,
    remote_schema TEXT
    )

### CODE REVIEW

File: scripts-available/CDB_FederatedServerTables.sql [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            'CREATE OR REPLACE VIEW %1$I AS
+                SELECT %2s
+                FROM %3$s t',
+            local_name,
+            array_to_string(carto_columns_expression || rest_of_cols, ','),
+            src_table
+        );
+    EXCEPTION
+    WHEN insufficient_privilege THEN
+        RAISE EXCEPTION 'Could not import table "%" as "%": "%" already exists', remote_table, local_name, local_name;
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerTables.sql]: Is "already exists" an accurate error message `WHEN insufficient_privilege`?

File: scripts-available/CDB_FederatedServer.sql [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+AS $$
+    SELECT right(internal_server_name,
+            char_length(internal_server_name::TEXT) - char_length(@extschema@.__CDB_FS_Name_Pattern()))::TEXT;
+$$
+LANGUAGE SQL IMMUTABLE PARALLEL SAFE;
+
+--
+-- Produce a valid name for a schema generated for the Federated Server 
+--
+CREATE OR REPLACE FUNCTION @extschema@.__CDB_FS_Generate_Schema_Name(internal_server_name TEXT, schema_name TEXT)
```
[the remark this answers is not in the document]
Comment (Algunenano) [reply scripts-available/CDB_FederatedServer.sql]: It should always be a NAME (as it's for internal use only). I'm fixing it.

File: scripts-available/CDB_FederatedServer.sql [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+
+--
+-- Grant access to a server
+--
+CREATE OR REPLACE FUNCTION @extschema@.CDB_Federated_Server_Grant_Access(server TEXT, db_role NAME)
+RETURNS void
+AS $$
+DECLARE
+    server_internal text := @extschema@.__CDB_FS_Generate_Server_Name(input_name := server, check_existence := true);
```
[the remark this answers is not in the document]
Comment (Algunenano) [reply scripts-available/CDB_FederatedServer.sql]: Done

File: scripts-available/CDB_FederatedServerTables.sql (hunk 1 above)
Comment (Algunenano) [reply scripts-available/CDB_FederatedServerTables.sql]: In this case, yes. I'm not sure why Postgres returns `insufficient_privilege` when creating a view conflicts with an existing table/view, but it does.

File: scripts-available/CDB_FederatedServer.sql [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+        END IF;
+    END LOOP;
+END
+$$
+LANGUAGE PLPGSQL VOLATILE PARALLEL UNSAFE;
+
+--
+-- Drops a registered server and all the objects associated with it
+-- 
+CREATE OR REPLACE FUNCTION @extschema@.CDB_Federated_Server_Unregister(server TEXT)
```
[the remark this answers is not in the document]
Comment (Algunenano) [reply scripts-available/CDB_FederatedServer.sql]: This affects both `CDB_Federated_Server_Unregister` and `CDB_Federated_Table_Unregister` directly and `CDB_Federated_Table_Register` indirectly (since it calls `CDB_Federated_Table_Unregister`), so I'm going to create an issue with this suggestion since adding it involves also adding quite a bit of tests.

File: scripts-available/CDB_FederatedServerListRemote.sql [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    -- "The information schema is defined in the SQL standard and can
+    -- therefore be expected to be portable and remain stable"
+    -- See https://www.postgresql.org/docs/11/information-schema.html
+    inf_schema name := 'information_schema';
+    remote_table name := 'schemata';
+    local_schema name := @extschema@.__CDB_FS_Create_Schema(server_internal, inf_schema);
+    role_name name := @extschema@.__CDB_FS_Generate_Server_Role_Name(server_internal);
+BEGIN
+    -- Import the foreign schemata table
+    IF NOT EXISTS (
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerListRemote.sql]: I think there's a common pattern here that could be refactored. In pseudo code: `IF NOT EXISTS table/schema THEN CREATE/IMPORT table/schema`.

BTW: you may want to check the target schema exists before

File: scripts-available/CDB_FederatedServer.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (rafatower) [reply scripts-available/CDB_FederatedServer.sql]: Alright. Please cc/ relevant people in the new ticket. We can discuss whether/when to implement it in the issue.

[the remark this answers is not in the document]
Comment (Algunenano) [reply scripts-available/CDB_FederatedServer.sql]: It's here: https://github.com/CartoDB/cartodb-platform/issues/6034

File: scripts-available/CDB_FederatedServerListRemote.sql (hunk 5 above)
Comment (Algunenano) [reply scripts-available/CDB_FederatedServerListRemote.sql]: The target schema is always created, if it doesn't already, exist with:
```
local_schema name := @extschema@.__CDB_FS_Create_Schema(server_internal, inf_schema);
```

File: scripts-available/CDB_FederatedServerDiagnostics.sql [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    local_schema name := @extschema@.__CDB_FS_Create_Schema(server_internal, remote_schema);
+    remote_server_version text;
+BEGIN
+    PERFORM @extschema@.__CDB_FS_Import_If_Not_Exists(server_internal, remote_schema, remote_table);
+
+    BEGIN
+        EXECUTE format('
+            SELECT setting FROM %I.%I WHERE name = ''server_version'';
+        ', local_schema, remote_table) INTO remote_server_version;
+    EXCEPTION WHEN OTHERS THEN
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerDiagnostics.sql]: may this be `WHEN insufficient_privilege`?

Comment (rafatower) [reply scripts-available/CDB_FederatedServerDiagnostics.sql]: I'd leave it "as is" for now, cause there may be other errors types (on the fdw category)

File: scripts-available/CDB_FederatedServerListRemote.sql [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    -- by its appropiate geometry and srid
+    RETURN QUERY EXECUTE format($q$
+        SELECT 
+            a.table_name::name,
+            a.column_name::name,
+            COALESCE(b.column_type, a.data_type)::TEXT as column_type
+        FROM
+            %I.%I a
+        LEFT JOIN
+            @extschema@.__CDB_FS_List_Foreign_Geometry_Columns_PG('%s', '%s') b
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerListRemote.sql]: just curious, how are they reported in the information schema?

File: scripts-available/CDB_FederatedServerListRemote.sql [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    END CASE;
+END
+$$
+LANGUAGE PLPGSQL VOLATILE PARALLEL UNSAFE;
+
+--
+-- List remote tables in a federated server that the current user has access to.
+-- For registered tables it returns also the associated configuration
+--
+CREATE OR REPLACE FUNCTION @extschema@.CDB_Federated_Server_List_Remote_Tables(server TEXT, remote_schema TEXT)
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerListRemote.sql]: :clap: :clap: :clap:

File: scripts-available/CDB_FederatedServerTables.sql [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+AS $$
+    WITH column_definitions AS
+    (
+        SELECT regexp_split_to_array(trim(leading from regexp_split_to_table(view_def, '\n')), ' ') AS col_def
+    )
+    SELECT trim(trailing ',' FROM split_part(col_def[2], '.', 2))
+    FROM column_definitions
+    WHERE trim(trailing ',' FROM col_def[array_length(col_def, 1)]) IN ('t.cartodb_id', 'cartodb_id')
+    LIMIT 1;
+$$ LANGUAGE SQL IMMUTABLE PARALLEL SAFE;
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerTables.sql]: this is not guaranteed to be immutable, is it?

File: scripts-available/CDB_FederatedServerTables.sql [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+    BEGIN
+        EXECUTE FORMAT('IMPORT FOREIGN SCHEMA %I LIMIT TO (%I) FROM SERVER %I INTO %I;',
+                        remote_schema, remote_table, server_internal, local_schema);
+    EXCEPTION WHEN OTHERS THEN
+        RAISE EXCEPTION 'Could not import schema "%" of server "%": %', remote_schema, server, SQLERRM;
+    END;
+    
+    BEGIN
+        src_table := format('%I.%I', local_schema, remote_table);
+    EXCEPTION WHEN OTHERS THEN
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerTables.sql]: weird place for an exception to happen (guess this happened during testing)

File: scripts-available/CDB_FederatedServerTables.sql [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+            'CREATE OR REPLACE VIEW %1$I.%2$I AS
+                SELECT %3s
+                FROM %4$s t',
+            server_internal, local_name,
+            array_to_string(carto_columns_expression || rest_of_cols, ','),
+            src_table
+        );
+    EXCEPTION WHEN OTHERS THEN
+        IF EXISTS (SELECT to_regclass(format('%1$I.%2$I', server_internal, local_name))) THEN
+            RAISE EXCEPTION 'Could not import table "%" as "%.%" already exists', remote_table, server_internal, local_name;
```
Comment (rafatower) [inline scripts-available/CDB_FederatedServerTables.sql]: `WHEN duplicate_table` ?

File: scripts-available/CDB_FederatedServerTables.sql (hunk 10 above)
Comment (Algunenano) [reply scripts-available/CDB_FederatedServerTables.sql]: Yes, casting `'%I.%I'` to REGCLASS will raise an exception if the object doesn't exist.

File: scripts-available/CDB_FederatedServerListRemote.sql (hunk 7 above)
Comment (Algunenano) [reply scripts-available/CDB_FederatedServerListRemote.sql]: As `USER-DEFINED`

File: scripts-available/CDB_FederatedServerTables.sql (hunk 9 above)
Comment (Algunenano) [reply scripts-available/CDB_FederatedServerTables.sql]: It is immutable because it receives a `TEXT` so it'll always do the same.

File: scripts-available/CDB_FederatedServerTables.sql (hunk 11 above)
Comment (Algunenano) [reply scripts-available/CDB_FederatedServerTables.sql]: Nope, in the case of the tests it's raising `INSUFFICIENT PRIVILEGE` so it's more complex than it seems.

**Comment by Algunenano** [comment]
> Hope listings have some ORDER BY somewhere to avoid testability issues.

All listing functions use `ORDER BY` so it's covered.
